---
title: "miSeq PICRUSt2 Data Analysis"
author: "Daniel Manter"
date: '2022-05-19'
output: html_document
params:
  # project info
  proj_dir: "/home/manterd/SHAI"
  proj_name: "AgroEco"
  query_fasta: "AgroEco.dada.rep_seqs.fasta"
  query_shared: "AgroEco.dada.tx.shared"

  # output directory
  out_dir: "AgroEco.picrust2"

  # program paths
  CONDA_PATH: "/home/manterd/anaconda3"
---

### Description:

> R markdown script for PICRUSt2 analysis of MiSeq data pre-processed with DADA2

### Required programs:

-   [Anaconda](https://www.anaconda.com)
-   [PISCRUSt2 v2.5.0](https://github.com/picrust/picrust2)
    -   installed in picrust2 virtual environment
-   [mothur v1.48](https://github.com/mothur/mothur)

### Project directory structure (\*denotes a required input file/folder)\`

        +-- proj_dir*
        |   +-- miSeq PICRUSt2.Rmd*            <-- (this file)
        |   +-- dada.rep_seqs.fasta*           <-- representative sequences from DADA2
        |   +-- dada.tx.shared*                <-- mothur formated otu table from DADA2
        |   +-- GIBBs.KO.numbers.csv*          <-- list of KEGG numbers to subset
        |   +-- current_files.summary          <-- files generated by mothur
        |   +-- <proj_name>.GIBBs.phyloseq.RDS <-- <u>final</u> output in phyloseq format
        |   +-- picrust2_out                   <-- output directory with picrust results
        |   |   +-- out.tre 
        |   |   +-- marker_predicted_and_nsti.tsv.gz 
        |   |   +-- KO_predicted.tsv.gz
        |   |   +-- intermediate
        |   |   |   +-- places_seqs
        |   |   +-- KO_metagenome_out
        |   |   |   +-- pred_metagenome_contrib.tsv.gz
        |   |   |   +-- pred_metagenome_unstrat.tsv.gz
        |   |   |   +-- gibbs_subset.tsv       <-- user-defined sub-set of KO's

### R setup

1.  Change python environment as needed

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, engine.opts = list(bash = "-l"))
Sys.setenv(RETICULATE_PYTHON="~/anaconda3/envs/picrust2/bin/python3.8")

# set parameters
for (key in names(params)) {
  do.call('Sys.setenv', params[key])
}

# Install and load necessary packages
# you may need permission to folders in .libPaths()

# install cran packages as needed
.cran_packages <- c("dplyr", "readxl", "tidyverse", "nlme")
new.packages <- .cran_packages[!(.cran_packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages, dependencies=TRUE, INSTALL_opts = c('--no-lock'))

# install BiocManager as needed 
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.15")

# install Bioconductor packages as needed
.bioc_packages <- c("phyloseq")
new.packages <- .bioc_packages[!(.bioc_packages %in% installed.packages()[,'Package'])]
if (length(new.packages)) BiocManager::install(new.packages)

# load all packages
sapply(c(.cran_packages, .bioc_packages), require, character.only=TRUE)

# set working directory
setwd(params$proj_dir)
```

### Convert shared file to biom file

```{bash mothur}
mothur "#DATA_LOCATION=$proj_dir; make.biom(shared=$query_shared); get.current()"
```

### Run PICRUSt2

-   no changes should be necessary

```{bash picrust2}
#mkdir $proj_dir/$out_dir
#mkdir $proj_dir/$out_dir/EC_metagenome_out

# activate conda environment where emu was installed
eval "$($CONDA_PATH/bin/conda shell.bash hook)"
conda activate picrust2

#place_seqs.py \
#  -s $proj_dir/$query_fasta \
#  -o $proj_dir/$out_dir/out.tre \
#  -p 1 \
#  --intermediate $proj_dir/$out_dir/intermediate/place_seqs

#hsp.py \
#  -i 16S \
#  -t $proj_dir/$out_dir/out.tre \
#  -o $proj_dir/$out_dir/marker_predicted_and_nsti.tsv.gz \
#  -p 1 -n

#hsp.py \
#  -i EC \
#  -t $proj_dir/$out_dir/out.tre \
#  -o $proj_dir/$out_dir/EC_predicted.tsv.gz \
#  -p 1

declare -- biom=$(sed -n "2p" $proj_dir/current_files.summary | cut -c 6-)
metagenome_pipeline.py \
  -i $proj_dir/$biom \
  -m $proj_dir/$out_dir/marker_predicted_and_nsti.tsv.gz \
  -f $proj_dir/$out_dir/EC_predicted.tsv.gz \
  -o $proj_dir/$out_dir/EC_metagenome_out \
  --strat_out 
```

### Select only EC's of interest

```{bash gibbs}
# If you change this list, you must update the GIBBs.KO.numbers.csv file
myList='EC.4.1.74\|EC.3.5.99.7\|EC.5.4.4.2\|EC.3.2.1.21\|EC.3.2.1.91\|EC.3.5.1.4\|EC.3.5.1.5\|EC.3.2.1.14\|EC.3.4.11.1\|EC.1.3.3.11\|EC.3.1.3.1\|EC.3.1.6.1\|EC.1.18.6.1\|EC.1.14.18.3\|EC.1.14.99.39\|EC.1.7.2.6\|EC.1.7.1.15\|EC.1.7.2.2\|EC.1.7.1.4\|EC.1.7.7.1\|EC.1.7.2.1\|EC.1.7.2.5\|EC.1.7.2.4'

myDir=$proj_dir/$out_dir/EC_metagenome_out
zcat $myDir/pred_metagenome_contrib.tsv.gz | head -n 1 > $myDir/gibbs_subset.tsv
zcat $myDir/pred_metagenome_contrib.tsv.gz | grep $myList >> $myDir/gibbs_subset.tsv
```

### Create phyloseq object with GIBBs data

```{r phyloseq_taxa}
library(dplyr)
library(phyloseq)
library(tidyr)

# get GIBBs info from file
tax <- read.csv(paste0(params$proj_dir, '/GIBBs.EC.numbers.csv'), sep=',')
TAX <- tax_table(as.matrix(tax))
taxa_names(TAX) <- tax$EC

# get GIBBs data from picrust2
file <- paste0(params$proj_dir, '/', params$out_dir, '/EC_metagenome_out/gibbs_subset.tsv')
func <- read.delim(file, sep='\t')
func <- func %>%
  group_by(sample, function.) %>%
  summarize(sum=sum(taxon_abun))
func <- func %>%
  spread(key=sample, value=sum)
func <- data.frame(func)
row.names(func) <- func$function.
func <- func[,-1]
func[is.na(func)] <- 0
OTU <- otu_table(func, taxa_are_rows=T)
row.names(OTU) <- gsub(":", ".", row.names(OTU))

#meta <- data.frame(read_excel(meta_file, sheet=sheet, range=range))
#common <- Reduce(intersect, list(meta$sample, names(otu)))
#meta <- meta[meta$sample %in% common,]
#SAMP <- sample_data(meta)
#sample_names(SAMP) <- meta[,1]
  
phy_obj <- phyloseq(OTU, TAX)
saveRDS(phy_obj, paste0(params$proj_dir, '/', params$out_dir, '.GIBBs.phyloseq.RDS'))
```
